src/main/preload/externalAPI.ts

```diff
@@ -0,0 +1,459 @@
+
+    // Utility
+    unregister: (channel) => ipcRenderer.removeAllListeners(channel),
+} as DesktopAPI);
+
+// Specific info for the testing environment
+if (process.env.NODE_ENV === 'test') {
+    contextBridge.exposeInMainWorld('testHelper', {
+        getViewInfoForTest: () => ipcRenderer.invoke(GET_VIEW_INFO_FOR_TEST),
+    });
+}
+
+/****************************************************************************
+ * window/document listeners
+ * These are here to perform specific tasks when global window or document events happen
+ * Avoid using these unless absolutely necessary
+ ****************************************************************************
+ */
+
+// Let the main process know when the window has finished resizing
+// This is to reduce the amount of white box that happens when expand the BrowserView
+window.addEventListener('resize', () => {
+    ipcRenderer.send(VIEW_FINISHED_RESIZING);
+});
+
+// Enable secure input on macOS clients when the user is on a password input
+let isPasswordBox = false;
+const shouldSecureInput = (element: {tagName?: string; type?: string} | null, force = false) => {
+    const targetIsPasswordBox = (element && element.tagName === 'INPUT' && element.type === 'password');
+    if (targetIsPasswordBox && (!isPasswordBox || force)) {
+        ipcRenderer.send(TOGGLE_SECURE_INPUT, true);
+    } else if (!targetIsPasswordBox && (isPasswordBox || force)) {
+        ipcRenderer.send(TOGGLE_SECURE_INPUT, false);
+    }
+
+    isPasswordBox = Boolean(targetIsPasswordBox);
+};
+window.addEventListener('focusin', (event) => {
+    shouldSecureInput(event.target as Element);
+});
+window.addEventListener('focus', () => {
+    shouldSecureInput(document.activeElement, true);
+});
+
+// exit fullscreen embedded elements like youtube - https://mattermost.atlassian.net/browse/MM-19226
+ipcRenderer.on('exit-fullscreen', () => {
+    if (document.fullscreenElement && document.fullscreenElement.nodeName.toLowerCase() === 'iframe') {
+        document.exitFullscreen();
+    }
+});
+
+// mattermost-webapp is SPA. So cache is not cleared due to no navigation.
+// We needed to manually clear cache to free memory in long-term-use.
+// http://seenaburns.com/debugging-electron-memory-usage/
+const CLEAR_CACHE_INTERVAL = 6 * 60 * 60 * 1000; // 6 hours
+setInterval(() => {
+    webFrame.clearCache();
+}, CLEAR_CACHE_INTERVAL);
+
+/****************************************************************************
+ * LEGACY CODE BELOW
+ * All of this code is deprecated and should be removed eventually
+ * Current it is there to support older versions of the web app
+ ****************************************************************************
+ */
+
+/**
+ * Legacy helper functions
+ */
+
+const onLoad = () => {
+    if (document.getElementById('root') === null) {
+        console.log('The guest is not assumed as mattermost-webapp');
```

Comment:
This is legacy code for one thing, and it was just a soft check to make sure we didn't try and scan the page or do anything to a page that isn't Mattermost, but I'll elaborate.

First, we don't actually support adding "servers" that aren't actually Mattermost servers. If a user decides to do that then they are responsible for any weird behaviours/consequences that come from doing that.

Second, this isn't actually a security check, it's really just to avoid any weird UX since we DID use to support having non-MM apps loaded into the app, but it didn't make sense to maintain. No matter what we do, there's always a way to have the inner client page access the Desktop App functionality (via just calling `window.postMessage` or calling an API method by name)

Lastly, the new API makes a different assumption than the old API did. The old API assumed the Web App was not (or sparsely aware) of the Desktop App and that the Desktop App had to do most of the heavy lifting in order to support its feature set. The new API assumes that the Web App is not only aware of the Desktop App, but will setup any communication between the two and send the required data needed. To summarize, Desktop App advertises what's available, and the Web App consumes it if needed.

Hope that kind of answers your question :P

Reply:
Yeah, I think so. If I understand right, we don't need to support the case any more that the page isn't part of MM, and we're doing things like setting the unread badge in a nicer way that wouldn't cause errors even if a non-MM app was loaded